lib/ansible/modules/network/panos/panos_dag_tags.py

```diff
@@ -0,0 +1,253 @@
+    commit:
+        description:
+            - commit if changed
+        required: false
+        default: true
+'''
+
+EXAMPLES = '''
+- name: Create the tags to map IP addresses
+  panos_dag_tags:
+    ip_address: "{{ mgmt_ip }}"
+    password: "{{ admin_password }}"
+    ip_to_register: "{{ ip_to_register }}"
+    tag_names: "{{ tag_names }}"
+    description: "Tags to allow certain IP's to access various SaaS Applications"
+    operation: 'add'
+  tags: "add-dagip"
+
+- name: List the IP address to tag mapping
+  panos_dag_tags:
+    ip_address: "{{ mgmt_ip }}"
+    password: "{{ admin_password }}"
+    tag_names: "{{ tag_names }}"
+    description: "List the IP address to tag mapping"
+    operation: 'list'
+  tags: "list-dagip"
+
+- name: Unregister an IP address from a tag mapping
+  panos_dag_tags:
+    ip_address: "{{ mgmt_ip }}"
+    password: "{{ admin_password }}"
+    ip_to_register: "{{ ip_to_register }}"
+    tag_names: "{{ tag_names }}"
+    description: "Unregister IP address from tag mappings"
+    operation: 'delete'
+  tags: "delete-dagip"
+'''
+
+RETURN = '''
+# Default return values
+'''
+
+
+from ansible.module_utils.basic import AnsibleModule, get_exception
+
+try:
+    from pandevice import base
+    from pandevice import firewall
+    from pandevice import panorama
+    from pandevice import objects
+
+    HAS_LIB = True
+except ImportError:
+    HAS_LIB = False
+
+
+def get_devicegroup(device, devicegroup):
+    dg_list = device.refresh_devices()
+    for group in dg_list:
+        if isinstance(group, panorama.DeviceGroup):
+            if group.name == devicegroup:
+                return group
+    return False
+
+
+def register_ip_to_tag_map(device, ip_addresses, tag):
+    """
+    :param device:
+    :param ip_addresses:
+    :param tag:
+    :return:
+    """
+
+    exc = None
+    try:
+        device.userid.register(ip_addresses, tag)
+    except Exception:
+        exc = get_exception()
+        return False, exc
+
+    if exc:
+        return False, exc
+    else:
+        return True, exc
+
+
+def get_all_address_group_mapping(device):
+    """
+    Retrieve all the tag to IP address mappings
+    :param device:
+    :return:
+    """
+
+    exc = None
+    ret = None
+    try:
+        ret = device.userid.get_registered_ip()
+    except Exception:
+        exc = get_exception()
+
+    if exc:
+        return False, exc
+    else:
+        return ret, exc
+
+
+def delete_address_from_mapping(device, ip_address, tags):
+    """
+    Delete an IP address from a tag mapping.
+    :param device:
+    :param ip_address:
+    :param tags:
+    :return:
+    """
+
+    exc = None
+    try:
+        ret = device.userid.unregister(ip_address, tags)
+    except Exception:
+        exc = get_exception()
+
+    if exc:
+        return False, exc
+    else:
+        return True, exc
+
+
+def main():
+    argument_spec = dict(
+        ip_address=dict(required=True),
+        password=dict(required=True),
```

Comment:
please add `,no_log=True`

Reply:
If you use an api_key is password still required?

You may wish to use `required_one_of=[['password', 'api_key']],` see https://github.com/gundalow/ansible/blob/docs-argspec/docs/docsite/rst/dev_guide/developing_modules_general.rst#main-and-ansiblemodule-argument-spec for more details

Reply:
OK, we used required_one before and this was oversight. Good catch and thank you.